File: tests/test-minecraft/src/main/java/pize/tests/minecraft/game/audio/AudioManager.java
package pize.tests.minecraft.game.audio;

import pize.Pize;
import pize.audio.sound.AudioSource;
import pize.app.Disposable;
import pize.tests.minecraft.game.Session;
import pize.tests.minecraft.game.options.SoundCategory;

public class AudioManager implements Disposable{

    public static final int MAX_SOUND_NUM = 128;

    private final Session session;
    private final AudioSource[] audioSources;


    public AudioManager(Session session){
        this.session = session;

        audioSources = new AudioSource[MAX_SOUND_NUM];
        for(int i = 0; i < audioSources.length; i++)
            audioSources[i] = new AudioSource();
    }


    public void play(Sound sound, float volume, float pitch){
        for(AudioSource audioSource: audioSources)
            if(!audioSource.isPlaying()){
                audioSource.setBuffer(session.getResourceManager().getSound(sound.resourceId));

                volume *= session.getOptions().getSoundVolume(sound.category) * sound.maxVolume;
                if(sound.category != SoundCategory.MASTER)
                    volume *= session.getOptions().getSoundVolume(SoundCategory.MASTER);

                audioSource.setVolume(volume);
                audioSource.setPitch(pitch);
                audioSource.play();

                return;
            }
    }

    public int getPlayingNum(){
        int playingSoundsNum = 0;
        for(AudioSource audioSource: audioSources)
            if(audioSource.isPlaying())
                playingSoundsNum++;
        return playingSoundsNum;
    }

    public void setDevice(String audioDevice){
        Pize.audio().getDevice(audioDevice).makeCurrent();
        for(AudioSource audioSource: audioSources)
            if(audioSource.isPlaying()){
                audioSource.pause();
                audioSource.play();
            }
    }


    @Override
    public void dispose(){
        for(AudioSource audioSource: audioSources)
            audioSource.dispose();
    }

}

File: tests/test-minecraft/src/main/kotlin/pize/tests/minecraft/game/audio/AudioManager.kt
package pize.tests.minecraft.game.audio

import pize.Pize.audio
import pize.app.Disposable
import pize.audio.sound.AudioSource
import pize.tests.minecraft.game.Session
import pize.tests.minecraft.game.options.SoundCategory

class AudioManager(private val session: Session) : Disposable {
    private val audioSources: Array<AudioSource?>

    init {
        audioSources = arrayOfNulls(MAX_SOUND_NUM)
        for (i in audioSources.indices) audioSources[i] = AudioSource()
    }

    fun play(sound: Sound, volume: Float, pitch: Float) {
        var volume = volume
        for (audioSource in audioSources) if (!audioSource!!.isPlaying) {
            audioSource.setBuffer(session.resourceManager.getSound(sound.resourceId))
            volume *= session.options.getSoundVolume(sound.category) * sound.maxVolume
            if (sound.category != SoundCategory.MASTER) volume *= session.options.getSoundVolume(SoundCategory.MASTER)
            audioSource.setVolume(volume.toDouble())
            audioSource.setPitch(pitch.toDouble())
            audioSource.play()
            return
        }
    }

    val playingNum: Int
        get() {
            var playingSoundsNum = 0
            for (audioSource in audioSources) if (audioSource!!.isPlaying) playingSoundsNum++
            return playingSoundsNum
        }

    fun setDevice(audioDevice: String?) {
        audio()!!.getDevice(audioDevice)!!.makeCurrent()
        for (audioSource in audioSources) if (audioSource!!.isPlaying) {
            audioSource.pause()
            audioSource.play()
        }
    }

    override fun dispose() {
        for (audioSource in audioSources) audioSource!!.dispose()
    }

    companion object {
        const val MAX_SOUND_NUM = 128
    }
}
